clarity/src/vm/analysis/type_checker/mod.rs

```diff
@@ -307,34 +371,311 @@ impl FunctionType {
         };
         check_argument_count(expected_args.len(), func_args)?;
 
-        for (expected_arg, arg) in expected_args.iter().zip(func_args.iter()).into_iter() {
-            match (&expected_arg.signature, arg) {
-                (
-                    TypeSignature::TraitReferenceType(trait_id),
-                    Value::Principal(PrincipalData::Contract(contract)),
-                ) => {
-                    let contract_to_check = db
-                        .load_contract(contract)
-                        .ok_or_else(|| CheckErrors::NoSuchContract(contract.name.to_string()))?;
-                    let trait_definition = db
-                        .get_defined_trait(&trait_id.contract_identifier, &trait_id.name)
-                        .unwrap()
-                        .ok_or(CheckErrors::NoSuchContract(
-                            trait_id.contract_identifier.to_string(),
-                        ))?;
-                    contract_to_check.check_trait_compliance(trait_id, &trait_definition)?;
+        let mut arg_types = Vec::new();
+        for arg in func_args {
+            arg_types.push(self.principal_to_callable_type(arg, 1)?);
+        }
+
+        for (expected_arg, arg_type) in expected_args.iter().zip(arg_types.iter()).into_iter() {
+            inner_type_check_type(
+                db,
+                None,
+                arg_type,
+                &expected_arg.signature,
+                clarity_version,
+                1,
+                &mut LimitedCostTracker::new_free(),
```

Comment:
Wait, why is type-checking a free operation here?

Reply:
Because this function is only called from `StacksChainState::can_include_tx`. I will add a comment to clarify this.

Reply:
69060afab